File: src/skin/Vertex.cpp
#include "java/util/ArrayList.txx"
#include "material/statistics.h"
#include "skin/Vertex.h"

unsigned int Vertex::currentComparisonFlags = VERTEX_COMPARE_LOCATION | VERTEX_COMPARE_NORMAL | VERTEX_COMPARE_TEXTURE_COORDINATE;

/**
Create a vertex with given coordinates, inNormal vector and list of inPatches
sharing the vertex. Several vertices can share the same coordinates
and inNormal vector. Several inPatches can share the same vertex
*/
Vertex::Vertex(
    Vector3D *inPoint,
    Vector3D *inNormal,
    Vector3D *inTextureCoordinates,
    java::ArrayList<Patch *> *inPatches):
    color(),
    tmp()
{
    id = GLOBAL_statistics.numberOfVertices++;
    point = inPoint;
    normal = inNormal;
    textureCoordinates = inTextureCoordinates;
    patches = inPatches;
    color.set(0.0f, 0.0f, 0.0f);
    radianceData = nullptr;
    back = (Vertex *)nullptr;
}

/**
Destroys the vertex. Does not destroy the coordinate vector and
normal vector, neither the patches sharing the vertex
*/
Vertex::~Vertex() {
    GLOBAL_statistics.numberOfVertices--;
    delete patches;
}

/**
Averages the color of each patch sharing the vertex and assign the 
resulting color to the vertex
*/
void
Vertex::computeColor() {
    long numberOfPatches;

    color.set(0.0f, 0.0f, 0.0f);
    numberOfPatches = 0;

    if ( patches != nullptr ) {
        for ( int i = 0; i < patches->size(); i++) {
            const Patch *patch = patches->get(i);
            color.r += patch->color.r;
            color.g += patch->color.g;
            color.b += patch->color.b;
        }
        numberOfPatches = patches->size();
    }

    if ( numberOfPatches > 0 ) {
        color.r /= (float) numberOfPatches;
        color.g /= (float) numberOfPatches;
        color.b /= (float) numberOfPatches;
    }
}

unsigned
Vertex::setCompareFlags(unsigned flags) {
    unsigned oldFlags = currentComparisonFlags;
    currentComparisonFlags = flags;
    return oldFlags;
}

File: typescript/src/main/vsdk/toolkit/skin/Vertex.ts
import { ColorRgb } from "../common/ColorRgb";
import { Vector3D } from "../common/linealAlgebra/Vector3D";
import { Statistics } from "../common/statistics/Statistics";
import { VertexCompareFlags } from "./VertexCompareFlags";
import type { Element } from "./Element";
import type { Patch } from "./Patch";

export class Vertex {
  private static currentComparisonFlags = VertexCompareFlags.VERTEX_COMPARE_LOCATION
    | VertexCompareFlags.VERTEX_COMPARE_NORMAL
    | VertexCompareFlags.VERTEX_COMPARE_TEXTURE_COORDINATE;

  public id: number;
  public point: Vector3D;
  public normal: Vector3D | null;
  public textureCoordinates: Vector3D | null;
  public color: ColorRgb;
  public radianceData: Element[] | null;
  public back: Vertex | null;
  public patches: Patch[] | null;
  public tmp: number;

  public constructor(
    inPoint: Vector3D,
    inNormal: Vector3D | null,
    inTextureCoordinates: Vector3D | null,
    inPatches: Patch[] | null
  ) {
    this.id = Statistics.instance().reader.numberOfVertices++;
    this.point = inPoint;
    this.normal = inNormal;
    this.textureCoordinates = inTextureCoordinates;
    this.patches = inPatches;
    this.color = new ColorRgb();
    this.color.set(0.0, 0.0, 0.0);
    this.radianceData = null;
    this.back = null;
    this.tmp = 0;
  }

  public destroy(): void {
    Statistics.instance().reader.numberOfVertices--;
    this.patches = null;
  }

  public computeColor(): void {
    this.color.set(0.0, 0.0, 0.0);
    let numberOfPatches = 0;

    if (this.patches !== null) {
      for (let i = 0; i < this.patches.length; i++) {
        const patch = this.patches[i];
        this.color.r += patch.color.r;
        this.color.g += patch.color.g;
        this.color.b += patch.color.b;
      }
      numberOfPatches = this.patches.length;
    }

    if (numberOfPatches > 0) {
      this.color.r /= numberOfPatches;
      this.color.g /= numberOfPatches;
      this.color.b /= numberOfPatches;
    }
  }

  public static setCompareFlags(flags: number): number {
    const oldFlags = Vertex.currentComparisonFlags;
    Vertex.currentComparisonFlags = flags;
    return oldFlags;
  }
}
